### PR TITLE
Mesh: Fix raycast for BufferGeometry and multi materials

### PR DESCRIPTION
**Pull request by Mugen87**
Resumes #11754

@WestLangley The processing of groups happens in `.raycast()` now. Do you think this approach is better compared to #11754?

### CODE REVIEW

**Comment by WestLangley** [comment]
@Mugen87 Yes, this is the right idea. I'll leave it up to you to verify, however. :)

Make sure your test case is simple enough to test manually, but not trivial -- you know, geometry indices are not in order, groups cover only part of the geometry, etc. There is also draw range...

File: src/objects/Mesh.js [hunk 1]
```diff
@@ -178,13 +178,13 @@ Mesh.prototype = Object.assign( Object.create( Object3D.prototype ), {
 
 		}
 
-		function checkBufferGeometryIntersection( object, raycaster, ray, position, uv, a, b, c ) {
+		function checkBufferGeometryIntersection( object, material, raycaster, ray, position, uv, a, b, c ) {
```
Comment (WestLangley) [inline src/objects/Mesh.js]: `raycaster` _and_ `ray`?

Comment (Mugen87) [inline src/objects/Mesh.js]: These parameters are pass through to `.checkIntersection()`.

Comment (Mugen87) [inline src/objects/Mesh.js]: Also see https://github.com/Mugen87/three.js/blob/82e866be2ffa752969af6efe96f91b5baab19c33/src/objects/Mesh.js#L230

Comment (Mugen87) [inline src/objects/Mesh.js]: Also see https://github.com/Mugen87/three.js/blob/82e866be2ffa752969af6efe96f91b5baab19c33/src/objects/Mesh.js#L230

**Comment by Mugen87** [comment]
I've tested the code with a modified example of `webgl_interactive_cubes.html`. Everything looks fine so far. Other examples with a "normal" raycast setup (without multi materials) work as before.

I'm still a little bit uncertain how to regard `drawRange` in context of groups. Is a combination basically possible?

**Comment by WestLangley** [comment]
Check how draw range is handled by the renderer. It is a bit complicated...

File: src/objects/Mesh.js (hunk 1 above)
Comment (WestLangley) [inline src/objects/Mesh.js]: Oh gosh. `ray` is in a different space?

Also, I see object, material, raycaster passed through to the lowest-level methods. Maybe you can refactor at some point... if you want...

Comment (Mugen87) [inline src/objects/Mesh.js]: > Maybe you can refactor at some point... if you want...

Not sure i wanna do this 😅. 

For this PR i'd like a solution with as few changes as possible.

**Comment by Mugen87** [comment]
related #8479

**Comment by Mugen87** [comment]
I've added support for `.drawRange` in `.raycast()`. This change also affects buffer geometries without multiple materials.

File: src/objects/Mesh.js [hunk 2]
```diff
@@ -245,24 +245,67 @@ Mesh.prototype = Object.assign( Object.create( Object3D.prototype ), {
...
-						intersection = checkBufferGeometryIntersection( this, raycaster, ray, position, uv, a, b, c );
+							group = groups[ i ];
+							groupMaterial = material[ group.materialIndex ];
 
-						if ( intersection ) {
+							for ( j = group.start, jl = ( group.start + group.count ); j < jl; j += 3 ) {
 
-							intersection.faceIndex = Math.floor( i / 3 ); // triangle number in indices buffer semantics
-							intersects.push( intersection );
+								if ( j >= drawStart && ( j + 2 ) < drawEnd ) {
```
Comment (WestLangley) [inline src/objects/Mesh.js]: As I mentioned earlier when discussing `drawRange`, I suggest you use the same coding pattern for raycasting that is used for rendering: https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L740-L751

Comment (Mugen87) [inline src/objects/Mesh.js]: I've looked into this but the situation in `.renderBufferDirect()` is different than in `.raycast()`. `.renderBufferDirect()` already gets the correct group and does not have to iterate over all existing groups like in `.raycast()`. This makes a direct comparison of both methods complicated. Besides, i'm not sure if we need to regard `rangeFactor` in `.raycast()`.

**Comment by Mugen87** [comment]
I've tried to adapt the coding pattern to `.raycast()`. Hope, it's okay like that 👍

**Comment by Mugen87** [comment]
BTW: `rangeFactor` seems to be wireframe specific and not relevant for the intersection test.

**Comment by WestLangley** [comment]
Nice! Thanks for working on this.

@mrdoob This looks good to me, but as I said, I'm leaving it to @Mugen87 to fully test.

**Comment by Mugen87** [comment]
I think.......it should work 😇

**Comment by Mugen87** [comment]
Also related #6939

**Comment by Mugen87** [comment]
An other issue in this context #11470

If there are concerns about the PR, maybe we can merge it at the beginning of the next dev cycle?

**Comment by hofk** [comment]
In my humble opinion, raycast for material arrays should be supported at the core of three. js.

Basic geometry such as BoxBufferGeometry, BoxGeometry use several materials.
https://github.com/mrdoob/three.js/blob/master/src/geometries/BoxGeometry.js

With this geometry it is easy to create a dice with dots, colors or images. Then you should be able to use raycast to select the cube pages correctly.

Regardless of that, I would appreciate it for my more complex figures if the feature is available in three.js.

**Comment by JohnRCook** [comment]
Is there a reason this fixed issue isn't promoted to a release yet? I have an issue with raycasting with double sided materials in a buffergeometry with groups.

**Comment by mybackstage** [comment]
Looking at the code, for indexed buffer geometry with multiple materials I think that the index to be used for calculation of the faceIndex should be  j instead of i (I tested it and it works for me), i.e.
intersection.faceIndex = Math.floor( j / 3 ); // triangle number in indices buffer semantics
instead of 
intersection.faceIndex = Math.floor( i / 3 ); // triangle number in indices buffer semantics

**Comment by mybackstage** [comment]
Looking at rel94 the problem is still there.
When a problem is found in a release, apart from fixing the master branch,
is a the problem also fixed in the release itself? (i.e. a new rel94 being
build?)

Thanks,
avner

**Comment by Mugen87** [comment]
No, there will be no new `R94` build (see https://github.com/mrdoob/three.js/pull/14392#issuecomment-401281106)

It will be fixed with the next release `R95`.

**Comment by mybackstage** [comment]
I guess my question is a general for release management, (beyond this
specific problem and comment).
If a problem is being found in a release is new release being created? Does
a release ever change after it is being created?
Are there guidelines for release management that you can refer me to?

I'm asking to know for the future. If I find a bug in a release, should I
expect it to get fixed? Or is it better to work with the master (or dev)
branch?
I originally worked with the master branch, but switched to a release
thinking that it will be more stable...

By the way, is there a IRC channel where such questions can be asked?

Thanks again!
Avner

**Comment by duhaime** [comment]
@Mugen87 Do you by chance have a sense of when R95 might be released?

**Pull request closed without merging**